poly-commitment/src/lagrange_cache.rs

```diff
@@ -0,0 +1,201 @@
+use std::{
+    collections::hash_map::DefaultHasher,
+    fs::{self, File},
+    marker::PhantomData,
+    path::{Path, PathBuf},
+};
+
+use crate::{commitment::CommitmentCurve, PolyComm};
+use ark_poly::Radix2EvaluationDomain as D;
+use std::hash::{Hash, Hasher};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum LagrangeCache<G> {
+    Local(FileCache<G>),
+    Remote(GoogleCloudCache<G>),
+}
+
+impl<G> Default for LagrangeCache<G> {
+    fn default() -> Self {
+        LagrangeCache::Local(FileCache::new(PathBuf::from("/tmp/lagrange_cache")))
+    }
+}
+
+// Keep this enum private for type safety
+pub enum LagrangeCacheKey {
+    Local(PathBuf),
+    Remote(()),
+}
+
+impl<G: CommitmentCurve> LagrangeCacheTrait<G> for LagrangeCache<G> {
+    type CacheKey = LagrangeCacheKey;
+
+    fn lagrange_basis_cache_key(&self, g: &Vec<G>, domain: &D<G::ScalarField>) -> Self::CacheKey {
+        match self {
+            LagrangeCache::Local(cache) => {
+                LagrangeCacheKey::Local(cache.lagrange_basis_cache_key(g, domain))
+            }
+            LagrangeCache::Remote(cache) => {
+                LagrangeCacheKey::Remote(cache.lagrange_basis_cache_key(g, domain))
+            }
+        }
+    }
+
+    fn load_lagrange_basis_from_cache(
+        &self,
+        g: &Vec<G>,
+        domain: &D<G::ScalarField>,
+    ) -> Option<Vec<PolyComm<G>>> {
+        match self {
+            LagrangeCache::Local(cache) => cache.load_lagrange_basis_from_cache(g, domain),
+            LagrangeCache::Remote(cache) => cache.load_lagrange_basis_from_cache(g, domain),
+        }
+    }
+
+    fn cache_lagrange_basis(
+        &self,
+        g: &Vec<G>,
+        domain: &D<G::ScalarField>,
+        basis: &Vec<PolyComm<G>>,
+    ) {
+        match self {
+            LagrangeCache::Local(cache) => cache.cache_lagrange_basis(g, domain, basis),
+            LagrangeCache::Remote(cache) => cache.cache_lagrange_basis(g, domain, basis),
+        }
+    }
+}
+
+pub trait LagrangeCacheTrait<G: CommitmentCurve> {
```

Comment:
It's generally considered bad practice/redundant to call a trait `Trait`

Reply:
Yes it's lame, thankfully don't need this trait anymore